packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -945,3 +1016,439 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+    required this.layoutSurrogate,
+    required Widget child,
+  }) : super(child: child);
+
+  final _RenderLayoutSurrogateProxyBox layoutSurrogate;
+
+  @override
+  _RenderDeferredLayoutBox createRenderObject(BuildContext context) {
+    final _RenderDeferredLayoutBox renderObject = _RenderDeferredLayoutBox(layoutSurrogate);
+    layoutSurrogate._deferredLayoutChild = renderObject;
+    return renderObject;
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderDeferredLayoutBox renderObject) {
+    assert(renderObject._layoutParent == layoutSurrogate);
+    layoutSurrogate._deferredLayoutChild = renderObject;
+  }
+}
+
+// A RenderProxyBox that defers its layout until its `_layoutParent` is laid out.
+//
+// This RenderObject must be a child RenderObject of a _RenderTheatre. It
+// guarantees that:
+//
+// 1. It's a relayout boundary. And it never dirties its parent _RenderTheatre
+//    or its _layoutParent.
+//
+// 2. Its `layout` implementation has been overridden such that `performLayout`
+//    does not do anything when its called from `layout`, preventing the parent
+//    `_RenderTheatre` from laying out this subtree prematurely (but this
+//    RenderObject may still be resized). Instead, `markNeedsLayout` will be
+//    called from within `layout` to schedule a layout update for this relayout
+//    boundary when needed.
+//
+// 3. When invoked from `PipelineOwner.flushLayout`, or
+//    _layoutParent.performLayout, this RenderObject behaves like a mini Overlay
+//    that has only one entry.
+class _RenderDeferredLayoutBox extends RenderProxyBox {
+  _RenderDeferredLayoutBox(this._layoutParent);
+
+  final _RenderLayoutSurrogateProxyBox _layoutParent;
+
+  @override
+  _RenderTheatre? get parent => super.parent as _RenderTheatre?;
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! StackParentData) {
+      child.parentData = StackParentData();
+    }
+  }
+
+  @override
+  void redepthChildren() {
+    _layoutParent.redepthChild(this);
+    super.redepthChildren();
+  }
+
+  bool _callingMarkParentNeedsLayout = false;
+  @override
+  void markParentNeedsLayout() {
+    // No re-entrant calls.
+    if (_callingMarkParentNeedsLayout) {
+      return;
+    }
+    _callingMarkParentNeedsLayout = true;
+    markNeedsLayout();
+    _layoutParent.markNeedsLayout();
+    _callingMarkParentNeedsLayout = false;
+  }
+
+  bool _needsLayout = true;
+  @override
+  void markNeedsLayout() {
+    _needsLayout = true;
+    super.markNeedsLayout();
+  }
+
+  @override
+  RenderObject? get debugLayoutParent => _layoutParent;
+
+  @override
+  bool get sizedByParent => true;
+
+  void layoutByLayoutParent() {
+    assert(!_parentDoingLayout);
+    final _RenderTheatre? theatre = parent;
+    if (theatre == null || !attached) {
+      assert(false, '$this is not attached to parent');
+      return;
+    }
+    super.layout(BoxConstraints.tight(theatre.constraints.biggest));
```

Comment:
layout surrogate always gets to do layout before `_RenderDeferredLayoutBox` so it's safer to get that from `theatre`, because `_constraints` could be null (or outdated) at this point I think.

Reply:
Hm, I guess it could be null. It's just odd that this class has information about this kind of implementation detail from the RenderTheatre...